app/lib/search/index_ducene.dart

```diff
@@ -0,0 +1,226 @@
+    if (query.type != null) {
+      duceneQuery
+          .addFilter(new TermQuery(new Term('detectedTypes', query.type)));
+    }
+
+    final int offset = query.offset ?? 0;
+    final int limit = max(minSearchLimit, query.limit ?? 100);
+
+    final TopDocs topPackages = await searcher.search(
+      duceneQuery,
+      maxSearchResults + minSearchLimit, // additional results for extended bias
+      scoreSort: new TFIDFScoreSort(),
+    );
+
+    List<PackageScore> packages = [];
+    for (ScoreDoc sd in topPackages.scoreDocs) {
+      final Document doc = await searcher.doc(sd.doc);
+      final double bias = double.parse(doc.get('popularity')) * 20;
+
+      packages.add(new PackageScore(
+        doc: new PackageDocument(
+          url: doc.get('id'),
+          package: doc.get('package'),
+          version: doc.get('version'),
+          devVersion: doc.get('devVersion'),
+          description: doc.get('originalDescription'),
+          lastUpdated: doc.get('lastUpdated'),
+          detectedTypes: doc.getValues('detectedTypes'),
+        ),
+        score: sd.score + bias,
+      ));
+    }
+    packages.sort((a, b) => -Comparable.compare(a.score, b.score));
+    if (packages.length > maxSearchResults) {
+      packages = packages.sublist(0, maxSearchResults);
+    }
+    final int totalCount = packages.length;
+
+    if (packages.length <= offset) {
+      packages = [];
+    } else {
+      packages = packages.sublist(offset);
+    }
+    if (packages.length > limit) {
+      packages = packages.sublist(0, limit);
+    }
+
+    return new PackageSearchResult(
+      lastUpdated: _lastUpdated.toIso8601String(),
+      totalCount: totalCount,
+      packages: packages,
+    );
+  }
+
+  String _normalizeWhitespaces(String text, int maxLength) {
+    if (text == null) return '';
+    String t = text.replaceAll(new RegExp('\\s+'), ' ').trim();
+    if (t.length > maxLength) {
+      t = t.substring(0, maxLength);
+    }
+    return t;
+  }
+
+  String _normalizeText(String text, int maxLength) {
+    if (text == null) return '';
+    String t = text
+        .toLowerCase()
+        .replaceAll(new RegExp('\\W'), ' ')
+        .replaceAll(new RegExp('\\s+'), ' ')
+        .trim();
+    if (t.length > maxLength) {
+      t = t.substring(0, maxLength);
+    }
+    return ' ' + t + ' ';
+  }
+}
+
+class _MultiNgramAnalyzer extends Analyzer {
+  _MultiNgramAnalyzer(int n) {
+    tokenizer = new _MultiNgramTokenizer(n);
+  }
+}
+
+class _MultiNgramTokenizer extends Tokenizer {
+  final int _n;
+  _MultiNgramTokenizer(this._n);
+
+  @override
+  Iterable<String> tokenize(String text) {
+    final Set<String> ngrams = new Set();
+    for (int ngramLength = 1; ngramLength <= _n; ngramLength++) {
+      if (text.length <= ngramLength) {
+        ngrams.add(text);
+      } else {
+        for (int i = 0; i <= text.length - ngramLength; i++) {
+          ngrams.add(text.substring(i, i + ngramLength));
```

Comment:
Yeah, it is the 1-gram, 2-gram, and 3-gram split of the string.
We may switch to words, but as long as we fit in memory easily, I don't mind having these around.